### ISSUE TITLE
Create better functionality for loss functions

### ISSUE THREAD

**Issue by abhmul** [abhmul/PyJet#6]
Currently loss functions and auxiliary loss functions are supported, but the setup is very clunky. One main loss function is supported, but the input must be stored in `self.loss_in` in the network `forward()` method. Auxiliary losses are appended to the `self.aux_loss_in` and these losses must be input into the `fit_generator` method in the exact same order their arguments were appended to `self.aux_loss_in`. Currently, assigning various optimizers to different losses is not supported.

Some avenues to improve functionality

* Use some kind of dictionary under the hood to link up losses by name with their inputs.
* Have some methods that abstract the use of this dictionary for users to use when creating network. Something like
```
add_loss_fn(self, name, loss_fn, *inputs)
```
* Store some kind of similar dictionary to figure out which optimizers go to which losses. This can be input into the fit_generator
* Consider making a `compile` method that will build this stuff before calling `fit_generator`. This will allow us to mimic keras better. This method `compile` will be able to to be called any number of times to change the loss + optimizer dictionaries.

**Comment by abhmul** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Create better functionality for loss functions](https://trello.com/c/nBNTZ7OS/9-create-better-functionality-for-loss-functions)
